File: app/src/main/java/com/olehliskovych/placenote/ui/base/BaseActivity.java
package com.olehliskovych.placenote.ui.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;

import com.olehliskovych.placenote.R;
import com.olehliskovych.placenote.databinding.ActivityMainBinding;

import javax.inject.Inject;

import dagger.android.support.DaggerAppCompatActivity;

public abstract class BaseActivity extends DaggerAppCompatActivity{

    private static final String FRAGMENT_KEY = "keyFragment";

    protected abstract void setupUI();
    public abstract ActivityMainBinding getBindng();

    protected BaseFragment currentFragment;

    @Inject
    FragmentManager fragmentManager;

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        final int lastIndexInFragmentStack = fragmentManager.getFragments().size() - 1;
        currentFragment = (BaseFragment) fragmentManager.getFragments().get(lastIndexInFragmentStack);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            restoreFragment(savedInstanceState);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (currentFragment != null) {
            String fragmentName = currentFragment.getClass().getName();
            fragmentManager.putFragment(outState, fragmentName, currentFragment);
        }
    }

    protected void restoreFragment(@NonNull Bundle savedInstanceState) {
        String fragmentName = savedInstanceState.getString(FRAGMENT_KEY, null);
        if (fragmentName != null) {
            currentFragment = (BaseFragment) fragmentManager.getFragment(savedInstanceState, fragmentName);
        }
    }

    public void navigateToFragment(@NonNull BaseFragment fragment,
                                   boolean isAddToBackStack,
                                   boolean isClearBackStack) {
        currentFragment = fragment;
        if (isClearBackStack) {
            clearBackStack();
        }
        if (isAddToBackStack) {
            fragmentManager.beginTransaction()
                    .replace(R.id.main_container, fragment)
                    .addToBackStack(fragment.getClass().getName())
                    .commit();
        } else {
            fragmentManager.beginTransaction()
                    .replace(R.id.main_container, fragment)
                    .commit();
        }
    }

    private void clearBackStack() {
        final int size = fragmentManager.getFragments().size() - 1;
        for (int i = 0; i <size; i++) {
            fragmentManager.popBackStack();
        }
    }

}

File: app/src/main/java/com/olehliskovych/placenote/ui/base/BaseActivity.kt
package com.olehliskovych.placenote.ui.base

import android.os.Bundle
import android.support.v4.app.FragmentManager
import com.olehliskovych.placenote.R
import dagger.android.support.DaggerAppCompatActivity
import javax.inject.Inject

abstract class BaseActivity : DaggerAppCompatActivity() {

    var currentFragment: BaseFragment? = null
    var currentFragmentName: String? = null

    companion object {
        private const val FRAGMENT_KEY = "keyFragment"
    }

    @Inject
    protected lateinit var mFragmentManager: FragmentManager

    protected abstract fun setupUI()

    override fun onBackPressed() {
        super.onBackPressed()
        val lastIndexInFragmentStack = mFragmentManager.fragments.size - 1
        currentFragment = mFragmentManager.fragments[lastIndexInFragmentStack] as BaseFragment
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState != null) {
            restoreFragment(savedInstanceState)
        }
    }

    protected enum class StatusBarColor {
        Master, Detail
    }

    protected fun setStatusBarColor(color: StatusBarColor) {
        if (color == StatusBarColor.Master) {

        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (currentFragment != null) {
            currentFragmentName = currentFragment!!.javaClass.name
            mFragmentManager.putFragment(outState, currentFragmentName, currentFragment)
        }
    }

    protected fun restoreFragment(savedInstanceState: Bundle) {
        currentFragmentName = savedInstanceState.getString(FRAGMENT_KEY, null)
        if (currentFragmentName != null) {
            currentFragment = mFragmentManager.getFragment(savedInstanceState, currentFragmentName) as BaseFragment
        }
    }

    fun navigateToFragment(fragment: BaseFragment,
                           isAddToBackStack: Boolean,
                           isClearBackStack: Boolean) {
        currentFragment = fragment
        if (isClearBackStack) {
            clearBackStack()
        }
        if (isAddToBackStack) {
            mFragmentManager.beginTransaction()
                    .replace(R.id.main_container, fragment)
                    .addToBackStack(fragment.javaClass.name)
                    .commit()
        } else {
            mFragmentManager.beginTransaction()
                    .replace(R.id.main_container, fragment)
                    .commit()
        }
    }

    private fun clearBackStack() {
        val size = mFragmentManager.fragments.size - 1
        for (i in 0 until size) {
            mFragmentManager.popBackStack()
        }
    }

}
































